pil_analyzer/src/evaluator.rs

```diff
@@ -1,104 +1,395 @@
+                        MatchPattern::CatchAll => Some(value),
                     })
-                    .expect("No arm matched the value {v}")
+                    .ok_or_else(EvalError::NoMatch)?;
+                evaluate(body, locals, symbols)?
             }
-            Expression::FreeInput(_) => Err("Cannot evaluate free input.".to_string()),
-        }
+            Expression::FreeInput(_) => Err(EvalError::Unsupported(
+                "Cannot evaluate free input.".to_string(),
+            ))?,
+        })
+    }
+
+    fn evaluate_reference<'a, T: FieldElement, C: Custom>(
+        reference: &Reference,
+        locals: &[Rc<Value<'a, T, C>>],
+        symbols: &impl SymbolLookup<'a, T, C>,
+    ) -> Result<Value<'a, T, C>, EvalError> {
+        Ok(match reference {
+            Reference::LocalVar(i, _name) => (*locals[*i as usize]).clone(),
+            Reference::Poly(poly) => symbols.lookup(&poly.name)?,
+        })
+    }
+}
+
+#[cfg(test)]
+mod test {
+    use number::GoldilocksField;
+    use pretty_assertions::assert_eq;
+
+    use crate::analyze_string;
+
+    use super::*;
+
+    fn parse_and_evaluate_symbol(input: &str, symbol: &str) -> String {
+        let analyzed = analyze_string::<GoldilocksField>(input);
+        let Some(FunctionValueDefinition::Expression(symbol)) = &analyzed.definitions[symbol].1
+        else {
+            panic!()
+        };
+        evaluate::<_, NoCustom>(symbol, &Definitions(&analyzed.definitions))
+            .unwrap()
+            .to_string()
+    }
+
+    #[test]
+    pub fn arrays_and_strings() {
+        let src = r#"namespace Main(16);
+            let words = ["the", "quick", "brown", "fox"];
+            let translate = |w| match w {
+                "the" => "franz",
+                "quick" => "jagt",
+                "brown" => "mit",
+                "fox" => "dem",
+                _ => "?",
+            };
+            let extract = |arr, i| arr[i];
+            let map_array = |arr, f| [f(arr[0]), f(arr[1]), f(arr[2]), f(arr[3])];
+            let translated = map_array(words, translate);
+        "#;
+        let result = parse_and_evaluate_symbol(src, "Main.translated");
+        assert_eq!(result, r#"["franz", "jagt", "mit", "dem"]"#);
+    }
+
+    #[test]
+    pub fn fibonacci() {
+        let src = r#"namespace Main(16);
+            let fib = |i| match i {
+                0 => 0,
+                1 => 1,
+                _ => fib(i - 1) + fib(i - 2),
+            };
+            let result = fib(20);
+        "#;
+        assert_eq!(parse_and_evaluate_symbol(src, "result"), "6765".to_string());
+    }
+
+    #[test]
+    pub fn capturing() {
+        let src = r#"namespace Main(16);
+            let f = |n, g| match n { 99 => |i| n, 1 => g(3) };
+            let result = f(1, f(99, |x| x + 3000));
+        "#;
+        // If the lambda function that f(99, ...) returns does not properly capture
+        // the value of n in a closure, f(1, ...) would return 1.
```

Comment:
This comment sounds a bit off

Reply:
Yeah, its grammar is a bit convoluted. What about:
```
If the lambda function returned by the expression f(99, ...) does not properly capture
the value of n in a closure, then f(1, ...) would return 1.
```

Reply:
sounds good!